src/transformers/file_utils.py

```diff
@@ -1163,8 +1163,16 @@ def __getitem__(self, k):
     def __setattr__(self, name, value):
         if name in self.keys() and value is not None:
             # Don't call self.__setitem__ to avoid recursion errors
-            super().__setitem__(name, value)
-        super().__setattr__(name, value)
+            if is_tf_available() and isinstance(value, tf.Tensor):
+                if name not in ("hidden_states", "attentions") or (value.shape[0] != 0 and value.shape[0] is not None):
```

Comment:
Just looking at the comment, there is a mistake in the behavior: all keys can be optional, not just "hidden_states" or "attentions".

Reply:
What are the cases where other keys than `loss`, `attentions`, `hidden_states` or `use_cache`  are `None`?

Reply:
`use_cache` is in no `ModelOutput` that I know of.
If you look at `modeling_tf_outputs` and `modeling_outputs` (for the output types currently missing in TF because the corresponding models have not been ported), you can see all keys that can appear in the various modeling outputs, and apart form the `logits`, they're pretty much all optionals. The `Seq2Seq` models have a lot for instance.

Reply:
Ok, I was mostly looking at the TF outputs not at the PT ones, and I didn't know about for the Seq2Seq models.

Reply:
Is there a way to ignore specific values on the fly and not at the creation of the output object?

Reply:
That's thing: I don't get why you ignore those values at the creation of the output object. You should only rely on the detection of empty tensors/None vlaues, not the key names.

Reply:
How would you proceed?

Reply:
I don't understand why you have to test for those keys in the first place so I can't advise you on that. The API should rely on the assumption that all keys are optionals since almost all of them will be in most of the model outputs.

Reply:
Let's forget about these keys. In the ModelOutput how would you proceed to ignore tensors that have their first dimension to `0` or `None` on the fly (so everytime you read the output basically) and not at the creation of the output?

Reply:
This should be treated the same way `None` values are currently treated: not set in the inner dict. I think that's exactly what you're doing here so this is good. It's the part of the test that looks at `hiddens` and `attentions` keys I don't see as needed.

Note that you need the same test in the init at the line `if v is not None` toward the end. You should not have the test a second time below in the setattr however.

Reply:
Something like this?

```
class ModelOutput(OrderedDict):
    def __post_init__(self):
        .....
        if other_fields_are_none and not is_tensor(first_field):
            ....
        else:
            for field in class_fields:
                v = getattr(self, field.name)
                if v is not None:
                    if is_tf_available() and isinstance(value, tf.Tensor):
                        if value.shape[0] != 0 and value.shape[0] is not None:
                            self[field.name] = v
                    else:
                        self[field.name] = v
    ......
    def __setattr__(self, name, value):
        if name in self.keys() and value is not None:
            # Don't call self.__setitem__ to avoid recursion errors
            if is_tf_available() and isinstance(value, tf.Tensor):
                if value.shape[0] != 0 and value.shape[0] is not None:
                    super().__setitem__(name, value)
            else:
                super().__setitem__(name, value)
        if is_tf_available() and isinstance(value, tf.Tensor):
            if value.shape[0] != 0 and value.shape[0] is not None:
                super().__setattr__(name, value)
        else:
            super().__setattr__(name, value)
```

Reply:
Yes, I'd just remove the last 5 lines and always leave `super().__setattr__(name, value)`: we always set the attribute but only registers it in the inner dict if it's not None/empty tensor.

Reply:
Ok, then it doesn't work. I get an error:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "c:\users\snake\dev\transformers\src\transformers\modeling_tf_utils.py", line 825, in from_pretrained
    model(model.dummy_inputs, training=False)  # build the network with dummy inputs
  File "C:\Users\snake\miniconda3\envs\transformers\lib\site-packages\tensorflow\python\keras\engine\base_layer.py", line 985, in __call__
    outputs = call_fn(inputs, *args, **kwargs)
  File "c:\users\snake\dev\transformers\src\transformers\modeling_tf_bert.py", line 958, in call
    outputs = self.bert(
  File "C:\Users\snake\miniconda3\envs\transformers\lib\site-packages\tensorflow\python\keras\engine\base_layer.py", line 985, in __call__
    outputs = call_fn(inputs, *args, **kwargs)
  File "c:\users\snake\dev\transformers\src\transformers\modeling_tf_bert.py", line 761, in call
    sequence_output = encoder_outputs[0]
  File "c:\users\snake\dev\transformers\src\transformers\file_utils.py", line 1165, in __getitem__
    return self.to_tuple()[k]
IndexError: tuple index out of range
```

The reason is because in graph mode all the tensors have their first dim to None. Then no more values are added this was why I have added some exceptions with the key names 😢 Will try to find another solution.

Reply:
If you want to see by yourself more exactly what is the issue, run the following piece of code:

```
from transformers import TFBertModel
import tensorflow as tf

model = tf.function(TFBertModel.from_pretrained("bert-base-cased"))
res = model(tf.constant([[1,2,3,4]]), output_attentions=tf.constant(True))
```

In `res` you have to have `attentions=<a tensor>` and `hidden_states=None`. Now with

```
from transformers import TFBertModel
import tensorflow as tf

model = tf.function(TFBertModel.from_pretrained("bert-base-cased"))
res = model(tf.constant([[1,2,3,4]]))
```

In `res` you have to have `attentions=None` and `hidden_states=None`.